static_frame/test/property/test_getsizeof_total.py

```diff
@@ -0,0 +1,102 @@
+import typing as tp
+from sys import getsizeof
+
+from hypothesis import given
+
+from static_frame import Frame
+from static_frame import Index
+from static_frame import IndexGO
+from static_frame import Series
+from static_frame import TypeBlocks
+from static_frame.core.util import getsizeof_total
+from static_frame.test.property import strategies as sfst
+from static_frame.test.test_case import TestCase
+
+
+class TestUnit(TestCase):
+
+    @given(sfst.get_index())
+    def test_getsizeof_total_index(self, i: Index) -> None:
+        seen: tp.Set[int] = set()
+        self.assertEqual(getsizeof_total(i), sum((
+            getsizeof_total(i._map, seen=seen),
+            getsizeof_total(i._labels, seen=seen),
+            getsizeof_total(i._positions, seen=seen),
+            getsizeof_total(i._recache, seen=seen),
+            getsizeof_total(i._name, seen=seen),
+            getsizeof(i) if id(i) not in seen else 0
+        )))
+
+    @given(sfst.get_index_go())
+    def test_getsizeof_total_index_go(self, i: IndexGO) -> None:
+        seen: tp.Set[int] = set()
+        self.assertEqual(getsizeof_total(i), sum((
+            getsizeof_total(i._map, seen=seen),
+            getsizeof_total(i._labels, seen=seen),
+            getsizeof_total(i._positions, seen=seen),
+            getsizeof_total(i._recache, seen=seen),
+            getsizeof_total(i._name, seen=seen),
+            getsizeof_total(i._labels_mutable, seen=seen),
+            getsizeof_total(i._labels_mutable_dtype, seen=seen),
+            getsizeof_total(i._positions_mutable_count, seen=seen),
+            getsizeof(i) if id(i) not in seen else 0
+        )))
+
+    @given(sfst.get_series())
+    def test_getsizeof_total_series(self, s: Series) -> None:
+        seen: tp.Set[int] = set()
+        self.assertEqual(getsizeof_total(s), sum((
+            getsizeof_total(s.values, seen=seen),
+            getsizeof_total(s._index, seen=seen),
+            getsizeof_total(s._name, seen=seen),
+            getsizeof(s) if id(s) not in seen else 0
+        )))
+
+    @given(sfst.get_type_blocks())
+    def test_getsizeof_total_type_blocks(self, tb: TypeBlocks) -> None:
+        # FAILING (maybe this is OK?), caused by numpy object arrays returning odd results, see below
+        '''
+        >>> import numpy as np
+        >>> a = np.array([np.array([None, None, None]), np.array([None, None, None])])
+        >>>
+        >>> id(a[0])
+        140280363921536
+        >>> id(a[0])
+        140280363921536
+        >>>
+        >>> a[0]
+        array([None, None, None], dtype=object)
+        >>>
+        >>> id(a[0])
+        140280362690400
+        >>> id(a[0])
+        140280363921536
+        >>> id(a[0])
+        140280363921536
+        >>> id(a[0])
+        140280363921536
+        '''
```

Comment:
This is failing because this form of numpy object array doesn't always return an object with the same ID. When this fails, it's off by 96, the size of one of the numpy arrays. Annoyingly, it does not fail consistently. For the most part (all the unit tests at least), this is not a problem. The property tests that make this fail have a numpy object array that contains multiple numpy object arrays of None (as shown in line 60).

Need to find out if there is a consistent way to get the contents of a numpy array for size counting. Alternatively, maybe the inconsistency is actually accurate to the memory that numpy is using?

Reply:
That is very surprising!

Reply:
I'm not sure if this is relevant here, as the dtype of `a` is object, but this answer suggests that because numpy does not store its array elements as python objects, it may create a new 'wrapper' python object on each access. I think this explanation is plausible if numpy uses a special placeholder for `None`. 

https://stackoverflow.com/a/39355706/1286571

Reply:
I'm removing a comment I had in code but removed. It may be useful to understand what I did to alleviate this problem.

```
        # The getsizeof_total function does not consistently size numpy object arrays since
        # numpy object arrays of numpy object arrays do not always return consistent elements.
        # While the ids of each element of a are the same, they may appear different during iteration.
        # For the purposes of only counting each unique element once, 
        #elements with unique ids, we cannot loop over each element in the numpy array to collect a list of unique objects
        '''
        >>> import numpy as np
        >>> a = np.array([np.array([None, None, None]) for i in range(5)])
        >>>
        >>> id(a[0]), id(a[1]), id(a[2]), id(a[3]), id(a[4])
        (140036333154352, 140036333154352, 140036333154352, 140036333154352, 140036333154352)
        >>> [id(n) for n in a]
        [140036333154352, 140036331924176, 140036333154352, 140036331924176, 140036333154352]
        '''
        '''
        >>> import numpy as np
        >>> a = np.array([np.array([None, None, i]) for i in range(5)])
        >>> b = [el for el in a]
        >>> c = [id(el) for el in a]
        >>>
        >>> a
        array([[None, None, 0],
               [None, None, 1],
               [None, None, 2],
               [None, None, 3],
               [None, None, 4]], dtype=object)
        >>> b
        [array([None, None, 0], dtype=object), array([None, None, 1], dtype=object), array([None, None, 2], dtype=object), array([None, None, 3], dtype=object), array([None, None, 4], dtype=object)]
        >>> c
        [139668158326624, 139668158326544, 139668158326624, 139668158326544, 139668158326624]
        >>> set(c)
        {139668158326624, 139668158326544}
        '''
```

Reply:
Figured out that by serializing each numpy object array element to a list, they will no longer have the duplicate id problem. See also: comment in `_unsized_children`.

Reply:
Even better, looping exclusively on the object references. See util.py:3202.